src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorMapToDocumentRangesEndpoint.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal class RazorMapToDocumentRangesEndpoint : IRazorMapToDocumentRangesHandler
+    {
+        private readonly RazorDocumentMappingService _documentMappingService;
+
+        public RazorMapToDocumentRangesEndpoint(RazorDocumentMappingService documentMappingService)
+        {
+            _documentMappingService = documentMappingService;
+        }
+
+        public bool MutatesSolutionState => false;
+
+        public TextDocumentIdentifier GetTextDocumentIdentifier(RazorMapToDocumentRangesParams request)
+        {
+            return new TextDocumentIdentifier
+            {
+                Uri = request.RazorDocumentUri
+            };
+        }
+
+        public async Task<RazorMapToDocumentRangesResponse?> HandleRequestAsync(RazorMapToDocumentRangesParams request, RazorRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            requestContext.RequireDocumentContext();
+            var documentContext = requestContext.DocumentContext;
+
+            if (request.Kind != RazorLanguageKind.CSharp)
+            {
+                // All other non-C# requests map directly to where they are in the document.
+                return new RazorMapToDocumentRangesResponse()
+                {
+                    Ranges = request.ProjectedRanges,
+                    HostDocumentVersion = documentContext.Version,
+                };
+            }
+
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+            var ranges = new Range[request.ProjectedRanges.Length];
+            for (var i = 0; i < request.ProjectedRanges.Length; i++)
+            {
+                var projectedRange = request.ProjectedRanges[i];
+                if (codeDocument.IsUnsupported() ||
```

Comment:
lol, so glad we're doing this check inside the loop _just to be sure_.

Reply:
https://github.com/dotnet/razor-tooling/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageEndpoint.cs#L155

Reply:
Oh, I know this isn't your code.. sorry, sometimes I forget to write that, its just that for some of this its the first time i'm seeing the code. I should probably log follow-ups